### PR TITLE
PERF: avoid copy in concatenate_array_managers if reindex already copies

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
@jbrockmendel https://github.com/pandas-dev/pandas/pull/42797 added an unconditional copy, while if you reindex the arrays, you already are sure you have a copy. 

This gives a 20% improvement on some of the merge benchmarks.

### CODE REVIEW

File: pandas/core/internals/concat.py [hunk 1]
```diff
@@ -94,7 +99,7 @@ def _concatenate_array_managers(
         # concatting along the columns -> combine reindexed arrays in a single manager
         assert concat_axis == 0
         arrays = list(itertools.chain.from_iterable([mgr.arrays for mgr in mgrs]))
-        if copy:
+        if copy and axis1_needs_copy_this:
```
Comment (jbrockmendel) [inline pandas/core/internals/concat.py:102]: might be cleaner (and avoid copies in corner cases) to do this up in the for loop right before `mgrs.append(mgr)`

**Comment by jreback** [comment]
lgtm @jbrockmendel comment could be addressed here

File: pandas/core/internals/concat.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/internals/concat.py:102]: Thanks, good idea, and simplified it a bit (and just needing a dummy copy() added to NullArrayProxy)

File: pandas/core/internals/concat.py [hunk 2]
```diff
@@ -77,10 +77,16 @@ def _concatenate_array_managers(
     # reindex all arrays
     mgrs = []
     for mgr, indexers in mgrs_indexers:
+        axis1_needs_copy = True
```
Comment (jbrockmendel) [inline pandas/core/internals/concat.py:80]: nitpick again: elsewhere in this file we do basically the same thing with `made_copy = False`

**Pull request merged**